### PR TITLE
Enable pulsar client support VPC-topic mode based on GSv1 and GSv2

### PR DESCRIPTION
**Pull request by VanderChen**
Due to the requirement of GoalState V2, this PR should be reviewed and merged after PR #625.

### CODE REVIEW

**Comment by cj-chung** [comment]
@VanderChen The current PR seems has overlaps with PR#625. Please consolidate these two PRs as one PR.

**Comment by xieus** [comment]
@VanderChen Ping me on Slack or here in this ticket when this is ready to review.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
+package com.futurewei.alcor.dataplane.cache;
```
[the remark this answers is not in the document]
Comment (xieus) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java:1]: @VanderChen This remains missing.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                result = entry.getValue().getSubscribeMapping().keySet().contains(nodeId);
+            }
+        }
+        return result;
+    }
+
+    @DurationStatistics
+    public void addSubscribedNodeForVpcId(String nodeId, String vpcId, String key) throws Exception {
+        VpcTopicInfo vpcTopicInfo = vpcTopicInfoICache.get(vpcId);
+        vpcTopicInfo.getSubscribeMapping().put(nodeId, key);
```
[the remark this answers is not in the document]
Comment (xieus) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java:80]: @VanderChen How is the resolved?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        vpcTopicInfoICache.put(vpcId, vpcTopicInfo);
+    }
```
[the remark this answers is not in the document]
Comment (xieus) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java:82]: @VanderChen is this resolved?

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/PathManagerService.java [hunk 4]
```diff
@@ -16,26 +16,20 @@
...
-        return null;
+    public boolean isFastPath() {
+        return USE_GRPC;
     }
 
-    @Override
-    public List<String> sendGoalStates(List<UnicastGoalStateV2> unicastGoalStates, MulticastGoalStateV2 multicastGoalState) throws Exception {
-        return null;
+    public boolean isFastPath(PortEntity portEntity) {
+        return USE_GRPC;
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/PathManagerService.java:33]: This is a good abstraction.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java]: A db transaction is used here to solve this problem.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/group_node_mode/DataPlaneClientImpl.java [hunk 5]
```diff
@@ -1,19 +1,21 @@
...
+ *     to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ *     The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ * /
+ */
+package com.futurewei.alcor.dataplane.client.pulsar.group_node_mode;
```
[the remark this answers is not in the document]
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/group_node_mode/DataPlaneClientImpl.java:18]: The changes are removed. The license style has been unified.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java]: DB transactions are added in VpcTopicCache.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/VpcTopicCache.java:16]: MIT license statement is added.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/vpc_mode/DataPlaneClientImpl.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+//@Component
```
[the remark this answers is not in the document]
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/vpc_mode/DataPlaneClientImpl.java:43]: Legacy annotation is removed.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/vpc_mode/PulsarConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+//@ConditionalOnProperty(prefix = "mq", name = "type", havingValue = "pulsar")
```
[the remark this answers is not in the document]
Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/vpc_mode/PulsarConfiguration.java]: Legacy annotation is removed.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/entity/MulticastGoalStateV2.java [hunk 8]
```diff
@@ -65,6 +72,13 @@ public void setVpcIds(List<String> vpcIds) {
         this.vpcIds = vpcIds;
     }
 
+    public void addVpcId(String vpcId) {
+        if (this.vpcIds == null) {
+            vpcIds = new ArrayList<String>();
+        }
+        this.vpcIds.add(vpcId);
```
Comment (cj-chung) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/entity/MulticastGoalStateV2.java:79]: @VanderChen What if `vpcId` is null?

Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/entity/MulticastGoalStateV2.java:79]: Should I check the input before usage or throw an exception here? Similar to the set method, it does not seem to check the validity of the input in the entity class.

Comment (cj-chung) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/entity/MulticastGoalStateV2.java:79]: @VanderChen throw an exception is good enough for here and set method. You probably also need to handle the exception from the caller too.

Comment (VanderChen) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/entity/MulticastGoalStateV2.java:79]: @cj-chung VpcIdNotFound() exception is added.

**Comment by xieus** [comment]
@VanderChen Could you start a Jenkins workflow to make sure no regression is observed, and plus get @cj-chung's sign off?

**Comment by VanderChen** [comment]
> @VanderChen Could you start a Jenkins workflow to make sure no regression is observed, and plus get @cj-chung's sign off?

I'll try in Jenkins as soon as possible. However, Jenkins seems only to support gRPC now.

**Comment by cj-chung** [comment]
@VanderChen The `USE_GRPC` in your `PathManagerService` is set to `true`:
``` 
private static boolean USE_GRPC = true;
```
That's mean your code should use grpc channel and pass in the Jenkins test, but it's failed in the `add_router_interface`.

```
Attaching subnets to router
PUTing http request
http://localhost:9009/project/3dda2801-d675-4688-a63f-dcda8d327f50/routers/11112801-d675-4688-a63f-dcda8d327f50/add_router_interface {'subnet_id': '8182a4d4-ffff-4ece-b3f0-8d36e3d88001'}
PUT Failed for http://localhost:9009/project/3dda2801-d675-4688-a63f-dcda8d327f50/routers/11112801-d675-4688-a63f-dcda8d327f50/add_router_interface with error
<bound method Response.json of <Response [500]>>
ERROR 500 Server Error: Internal Server Error for url: http://localhost:9009/project/3dda2801-d675-4688-a63f-dcda8d327f50/routers/11112801-d675-4688-a63f-dcda8d327f50/add_router_interface
500 Server Error: Internal Server Error for url: http://localhost:9009/project/3dda2801-d675-4688-a63f-dcda8d327f50/routers/11112801-d675-4688-a63f-dcda8d327f50/add_router_interface
Build step 'Execute shell' marked build as failure
```

@pkommoju Please add more debug info here from Jenkins. Thanks.

**Comment by VanderChen** [comment]
> @VanderChen The `USE_GRPC` in your `PathManagerService` is set to `true`:
> 
> ```
> private static boolean USE_GRPC = true;
> ```
> 
> That's mean your code should use grpc channel and pass in the Jenkins test, but it's failed in the `add_router_interface`.
> 
> ```
> Attaching subnets to router
> PUTing http request
> http://localhost:9009/project/3dda2801-d675-4688-a63f-dcda8d327f50/routers/11112801-d675-4688-a63f-dcda8d327f50/add_router_interface {'subnet_id': '8182a4d4-ffff-4ece-b3f0-8d36e3d88001'}
> PUT Failed for http://localhost:9009/project/3dda2801-d675-4688-a63f-dcda8d327f50/routers/11112801-d675-4688-a63f-dcda8d327f50/add_router_interface with error
> <bound method Response.json of <Response [500]>>
> ERROR 500 Server Error: Internal Server Error for url: http://localhost:9009/project/3dda2801-d675-4688-a63f-dcda8d327f50/routers/11112801-d675-4688-a63f-dcda8d327f50/add_router_interface
> 500 Server Error: Internal Server Error for url: http://localhost:9009/project/3dda2801-d675-4688-a63f-dcda8d327f50/routers/11112801-d675-4688-a63f-dcda8d327f50/add_router_interface
> Build step 'Execute shell' marked build as failure
> ```
> 
> @pkommoju Please add more debug info here from Jenkins. Thanks.
> 
> *** No more debug info is possible in this context, the call is into a python module (requests.post), and on error all of the response is printed. Increasing the application logging level might help.

@cj-chung Hi, James. I have noted this when the test is run in Jenkins. I'm working on it locally to solve this 500 error. This fix will be submitted as soon as possible.

**Comment by cj-chung** [comment]
@VanderChen Which part of code you have modified to make the Jenkins test cases passed?

**Review by cj-chung — APPROVED** [review]
LGTM

**Pull request merged**